decksite/achievements.py

```diff
@@ -0,0 +1,237 @@
+from typing import TYPE_CHECKING, Dict, List, Optional
+
+from flask import url_for
+from flask_babel import ngettext
+
+import decksite
+from decksite.data import query
+from magic import tournaments
+
+if TYPE_CHECKING:
+    from decksite.data import person # pylint:disable=unused-import
+# Disabling unused-import supposedly not needed here but actually seems to be?
+
+def load_query(people_by_id: Dict[int, 'person.Person'], season_id: Optional[int]) -> str:
+    columns = ', '.join(f'SUM({a.key}) as {a.key}' for a in Achievement.all_achs if a.in_db)
+    return """
+        SELECT
+            person_id AS id,
+            {columns}
+        FROM
+            _achievements AS a
+        WHERE
+            person_id IN ({ids}) AND ({season_query})
+        GROUP BY
+            person_id
+    """.format(columns=columns, ids=', '.join(str(k) for k in people_by_id.keys()), season_query=query.season_query(season_id))
+
+def preaggregate_query() -> str:
+    create_columns = ', '.join(f'{a.key} INT NOT NULL' for a in Achievement.all_achs if a.in_db)
+    select_columns = ', '.join(f'{a.sql} as {a.key}' for a in Achievement.all_achs if a.in_db)
+    return """
+        CREATE TABLE IF NOT EXISTS _new_achievements (
+            person_id INT NOT NULL,
+            season_id INT NOT NULL,
+            {cc},
+            PRIMARY KEY (season_id, person_id),
+            FOREIGN KEY (season_id) REFERENCES season (id) ON UPDATE CASCADE ON DELETE CASCADE,
+            FOREIGN KEY (person_id) REFERENCES person (id) ON UPDATE CASCADE ON DELETE CASCADE
+        ) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci AS
+        SELECT
+            p.id AS person_id,
+            season.id AS season_id,
+            {sc}
+        FROM
+            person AS p
+        LEFT JOIN
+            deck AS d ON d.person_id = p.id
+        LEFT JOIN
+            deck_cache AS dc ON dc.deck_id = d.id
+        {season_join}
+        {competition_join}
+        GROUP BY
+            p.id,
+            season.id
+        HAVING
+            season.id IS NOT NULL
+    """.format(cc=create_columns, sc=select_columns, season_join=query.season_join(), competition_join=query.competition_join())
+
+def descriptions() -> List[Dict[str, str]]:
+    return [{'title': a.title, 'description_safe': a.description_safe} for a in Achievement.all_achs]
+
+def displayed_achievements(p: 'person.Person') -> List[Dict[str, str]]:
+    return [d for d in (a.display(p) for a in Achievement.all_achs) if d is not None]
+
+# Abstract achievement classes
+
+class Achievement:
```

Comment:
Cool stuff. I had not seen Python classes used like this before (like a struct, and with @property). I wonder if some of our models should use this style instead of being Containers.